### ISSUE TITLE
Invalid cross-device link

### ISSUE THREAD

**Issue by muk-ai** [toy/image_optim#180]
I'll leave a report for anyone who encounters the same error.

## error message
```
Invalid cross-device link @ rb_file_s_rename - (/tmp/fat20200909-9154-12oecct.png, /var/www/my-app-name/releases/20200909065727/tmp/1599636588-64606488395367-0002-2349/fat.png)
```

## related gems and configuration

- image_optim (0.27.0)
- carrierwave (2.1.0)
- rails
- passenger
- nginx
- systemd

systemd is configured to give nginx a private /tmp.
```
PrivateTmp=true
```

## why the error occurred

1. carrierwave saves the uploaded file to `/var/www/my-app-name/releases/20200909065727/tmp`. (it is `Rails.root.join('tmp')`)
2. image_optim tries to create a hard link in `/tmp` when processing the file.
3. Since `/var/www/my-app-name/releases/2020090909065727/tmp` and `/tmp` have different filesystem namespaces, "Invalid cross-device link" error occurs.

## our solution

Changed tmp directory path by specifying `ENV['TMPDIR']` in order to use the same tmp directory.
Or turn off the `PrivateTmp` feature.

**Comment by toy** [comment]
Can you please confirm that the bug was introduced in 0.27.0? Does it work without specifying `TMPDIR` using version 0.26.5?
Also is `/var/www/my-app-name/releases/20200909065727/tmp` or `/tmp` a symlink? I'm trying to understand if same device check is wrong.

**Comment by muk-ai** [comment]
First of all, thank you for maintaining this gem.

> Can you please confirm that the bug was introduced in 0.27.0?

An error occurred on 0.27.0.
We had no problems with 0.26.5.

> Also is /var/www/my-app-name/releases/20200909065727/tmp or /tmp a symlink?

They are not symlinks.
However, we  are using the following symlink to release app.
/var/www/my-app-name/current -> /var/www/my-app-name/releases/20200909065727

> Is it also possible to share a backtrace of the exception?

I hope this helps.
```
/GEM_ROOT/gems/image_optim-0.27.0/lib/image_optim/path.rb:56→ rename
/GEM_ROOT/gems/image_optim-0.27.0/lib/image_optim/path.rb:56→ rename
/GEM_ROOT/gems/image_optim-0.27.0/lib/image_optim/path.rb:56→ replace
/home/ec2-user/.rbenv/versions/2.7.1/lib/ruby/2.7.0/delegate.rb:343→ block in delegating_block
/GEM_ROOT/gems/image_optim-0.27.0/lib/image_optim.rb:133→ optimize_image!
/PROJECT_ROOT/app/uploaders/base_uploader.rb:24→ optimize
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/processing.rb:83→ block (2 levels) in process!
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/processing.rb:75→ each
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/processing.rb:75→ block in process!
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/callbacks.rb:15→ with_callbacks
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/processing.rb:74→ process!
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/callbacks.rb:14→ block in with_callbacks
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/callbacks.rb:14→ each
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/callbacks.rb:14→ with_callbacks
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/uploader/cache.rb:144→ cache!
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/mounter.rb:63→ block (2 levels) in cache
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/mounter.rb:176→ handle_error
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/mounter.rb:47→ block in cache
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/mounter.rb:46→ map
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/mounter.rb:46→ cache
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/mount.rb:146→ opus_image=
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/mount.rb:373→ opus_image=
/GEM_ROOT/gems/carrierwave-2.1.0/lib/carrierwave/orm/activerecord.rb:75→ opus_image=
/GEM_ROOT/gems/activemodel-6.0.3.3/lib/active_model/attribute_assignment.rb:50→ public_send
/GEM_ROOT/gems/activemodel-6.0.3.3/lib/active_model/attribute_assignment.rb:50→ _assign_attribute
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/attribute_assignment.rb:29→ block in assign_nested_parameter_attributes
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/attribute_assignment.rb:29→ each
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/attribute_assignment.rb:29→ assign_nested_parameter_attributes
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/attribute_assignment.rb:23→ _assign_attributes
/GEM_ROOT/gems/activemodel-6.0.3.3/lib/active_model/attribute_assignment.rb:35→ assign_attributes
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/core.rb:325→ initialize
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/inheritance.rb:70→ new
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/inheritance.rb:70→ new
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/reflection.rb:158→ build_association
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/associations/association.rb:321→ build_record
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/associations/collection_association.rb:108→ build
/GEM_ROOT/gems/activerecord-6.0.3.3/lib/active_record/associations/collection_proxy.rb:316→ build
/PROJECT_ROOT/app/api/vivivit_api/works.rb:110→ block (2 levels) in <class:Works>
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:59→ call
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:59→ block (2 levels) in generate_api_method
/GEM_ROOT/gems/activesupport-6.0.3.3/lib/active_support/notifications.rb:182→ instrument
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:58→ block in generate_api_method
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:341→ execute
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:267→ block in run
/GEM_ROOT/gems/activesupport-6.0.3.3/lib/active_support/notifications.rb:182→ instrument
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:247→ run
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:322→ block in build_stack
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/grape-1.4.0/lib/grape/middleware/base.rb:35→ call!
/GEM_ROOT/gems/grape-1.4.0/lib/grape/middleware/base.rb:28→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/grape-1.4.0/lib/grape/middleware/error.rb:39→ block in call!
/GEM_ROOT/gems/grape-1.4.0/lib/grape/middleware/error.rb:38→ catch
/GEM_ROOT/gems/grape-1.4.0/lib/grape/middleware/error.rb:38→ call!
/GEM_ROOT/gems/grape-1.4.0/lib/grape/middleware/base.rb:28→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/head.rb:12→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:231→ call!
/GEM_ROOT/gems/grape-1.4.0/lib/grape/endpoint.rb:225→ call
/GEM_ROOT/gems/grape-1.4.0/lib/grape/router/route.rb:58→ exec
/GEM_ROOT/gems/grape-1.4.0/lib/grape/router.rb:116→ process_route
/GEM_ROOT/gems/grape-1.4.0/lib/grape/router.rb:72→ block in identity
/GEM_ROOT/gems/grape-1.4.0/lib/grape/router.rb:91→ transaction
/GEM_ROOT/gems/grape-1.4.0/lib/grape/router.rb:70→ identity
/GEM_ROOT/gems/grape-1.4.0/lib/grape/router.rb:55→ block in call
/GEM_ROOT/gems/grape-1.4.0/lib/grape/router.rb:132→ with_optimization
/GEM_ROOT/gems/grape-1.4.0/lib/grape/router.rb:54→ call
/GEM_ROOT/gems/grape-1.4.0/lib/grape/api/instance.rb:167→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/grape.rb:125→ call_with_new_relic
/GEM_ROOT/gems/grape-1.4.0/lib/grape/api/instance.rb:71→ call!
/GEM_ROOT/gems/grape-1.4.0/lib/grape/api/instance.rb:66→ call
/GEM_ROOT/gems/grape-1.4.0/lib/grape/api.rb:68→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/routing/mapper.rb:19→ block in <class:Constraints>
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/routing/mapper.rb:48→ serve
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/journey/router.rb:49→ block in serve
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/journey/router.rb:32→ each
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/journey/router.rb:32→ serve
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/routing/route_set.rb:834→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/coverband-5.0.1/lib/coverband/integrations/background_middleware.rb:10→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/rack/agent_hooks.rb:30→ traced_call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/warden-1.2.8/lib/warden/manager.rb:36→ block in call
/GEM_ROOT/gems/warden-1.2.8/lib/warden/manager.rb:34→ catch
/GEM_ROOT/gems/warden-1.2.8/lib/warden/manager.rb:34→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/tempfile_reaper.rb:15→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/etag.rb:27→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/conditional_get.rb:40→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/head.rb:12→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/http/content_security_policy.rb:18→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/PROJECT_ROOT/lib/classes/vi_request_logger.rb:10→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:266→ context
/GEM_ROOT/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:260→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/cookies.rb:648→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/callbacks.rb:27→ block in call
/GEM_ROOT/gems/activesupport-6.0.3.3/lib/active_support/callbacks.rb:101→ run_callbacks
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/callbacks.rb:26→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/actionable_exceptions.rb:17→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/airbrake-11.0.0/lib/airbrake/rack/middleware.rb:34→ call!
/GEM_ROOT/gems/airbrake-11.0.0/lib/airbrake/rack/middleware.rb:23→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/debug_exceptions.rb:32→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/show_exceptions.rb:33→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/chewy-5.1.0/lib/chewy/railtie.rb:19→ block in call
/GEM_ROOT/gems/chewy-5.1.0/lib/chewy/strategy.rb:70→ wrap
/GEM_ROOT/gems/chewy-5.1.0/lib/chewy.rb:201→ strategy
/GEM_ROOT/gems/chewy-5.1.0/lib/chewy/railtie.rb:19→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/railties-6.0.3.3/lib/rails/rack/logger.rb:37→ call_app
/GEM_ROOT/gems/railties-6.0.3.3/lib/rails/rack/logger.rb:28→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/remote_ip.rb:81→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/request_id.rb:27→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/method_override.rb:24→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/runtime.rb:22→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/executor.rb:14→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/rack-2.2.3/lib/rack/sendfile.rb:110→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/actionpack-6.0.3.3/lib/action_dispatch/middleware/host_authorization.rb:76→ call
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/GEM_ROOT/gems/railties-6.0.3.3/lib/rails/engine.rb:527→ call
/GEM_ROOT/gems/railties-6.0.3.3/lib/rails/railtie.rb:190→ public_send
/GEM_ROOT/gems/railties-6.0.3.3/lib/rails/railtie.rb:190→ method_missing
/GEM_ROOT/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101→ call
/home/ec2-user/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/passenger-6.0.6/src/ruby_supportlib/phusion_passenger/rack/thread_handler_extension.rb:107→ process_request
/home/ec2-user/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/passenger-6.0.6/src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb:157→ accept_and_process_next_request
/home/ec2-user/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/passenger-6.0.6/src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb:110→ main_loop
/home/ec2-user/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/passenger-6.0.6/src/ruby_supportlib/phusion_passenger/request_handler.rb:415→ block (3 levels) in start_threads
/home/ec2-user/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/passenger-6.0.6/src/ruby_supportlib/phusion_passenger/utils.rb:113→ block in create_thread_and_abort_on_exception
```

**Comment by toy** [comment]
After few attempts to reproduce it using docker and vagrant I still can't find how "same device" check would be true for temp file path and destination directory.
For your original example following should be true `File.stat('/tmp/fat20200909-9154-12oecct.png').dev == File.stat('/var/www/my-app-name/releases/20200909065727/tmp/1599636588-64606488395367-0002-2349').dev`. But `/tmp` should already be on tempfs and `PrivateTmp=true` if I understood correctly mounts subfolder from that temporary file system as `/tmp` for the process, so devices should have different identifier.

Can you come up with reproducible example/environment?

**Comment by muk-ai** [comment]
We didn't use `/tmp` as tmpfs. (Amazon Linux 2 image)
```
$ LANG=C df -hT
Filesystem     Type      Size  Used Avail Use% Mounted on
devtmpfs       devtmpfs  1.9G     0  1.9G   0% /dev
tmpfs          tmpfs     1.9G     0  1.9G   0% /dev/shm
tmpfs          tmpfs     1.9G  408K  1.9G   1% /run
tmpfs          tmpfs     1.9G     0  1.9G   0% /sys/fs/cgroup
/dev/nvme0n1p1 xfs        40G   33G  7.6G  82% /
tmpfs          tmpfs     389M     0  389M   0% /run/user/1000
$
```


> Can you come up with reproducible example/environment?

It may take some time, but I try to prepare a reproduction environment.

**Comment by muk-ai** [comment]
I created a minimal reproducible example.
https://github.com/muk-ai/image_optim_namespace_repro
To be safe, please run it in a Linux virtual machine.

step 1. start the bash process in an isolated namespace.
```
$ sudo unshare --mount /bin/bash
```

step 2. create a temporary directory and mount it as /tmp.
```
# PRIVATE_TMPDIR=$(mktemp -d)
# mount --bind $PRIVATE_TMPDIR /tmp
```

step 3. create a hard link.
```
# LANG=C ln image.png /tmp/hardlink
ln: failed to create hard link '/tmp/hardlink' => 'image.png': Invalid cross-device link
```

step 4. make sure that the same error occurs with optimize_image.
```
# bundle exec ruby optimize.rb
WARN: advpng 1.15 at /bin/advpng (< 1.17) does not use zopfli
Traceback (most recent call last):
	5: from optimize.rb:12:in `<main>'
	4: from /usr/local/rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/image_optim-0.27.0/lib/image_optim.rb:133:in `optimize_image!'
	3: from /usr/local/rbenv/versions/2.7.1/lib/ruby/2.7.0/delegate.rb:343:in `block in delegating_block'
	2: from /usr/local/rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/image_optim-0.27.0/lib/image_optim/path.rb:56:in `replace'
	1: from /usr/local/rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/image_optim-0.27.0/lib/image_optim/path.rb:56:in `rename'
/usr/local/rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/image_optim-0.27.0/lib/image_optim/path.rb:56:in `rename': Invalid cross-device link @ rb_file_s_rename - (/tmp/image20200918-15809-1nfzntw.png, ./image.png) (Errno::EXDEV)
```

**Comment by toy** [comment]
Thanks for creating the example!

I was able to test that mounting directory from same file system will preserve device id but will create a boundary for creating links. This means that checking `File::Stat#dev` in ruby code will not be enough. Additionally looking at source of `mv` ([freebsd](https://github.com/freebsd/freebsd/blob/master/bin/mv/mv.c), [gnu mv](https://git.savannah.gnu.org/gitweb/?p=coreutils.git;a=blob;f=src/mv.c) -> [gnu copy](https://git.savannah.gnu.org/gitweb/?p=coreutils.git;a=blob;f=src/copy.c)) it looks like rescuing `Errno::EXDEV` and copying file is the right thing to do. I'm wondering if it is better also to show a warning or consider it a normal behaviour.

**Comment by muk-ai** [comment]
Thanks for also checking out the `mv` behavior.
I don't think a warning is necessary. Because `mv` does not show any warning.

**Comment by toy** [comment]
Can you please check if changes in [branch fix-180](https://github.com/toy/image_optim/compare/fix-180) fix the problem?

**Comment by muk-ai** [comment]
It works perfectly in `PrivateTmp` enabled environment :tada:
On behalf of our team, I would like to thank you for support of this edge case.

**Comment by toy** [comment]
Released version 0.27.1
